### PR TITLE
Add example support

### PR DESCRIPTION
**Pull request by fritzbrand**
## Purpose
WhatsApp made a change across versions of the API, that makes supplying example values mandatory, when using variables in a template

## Approach
Add support for example values in the ContentRepo, and add the functionality to include them in the template creating process

### CODE REVIEW

**Comment by fritzbrand** [comment]
@rudigiesler the last check is taking forever, so I am going to assume it's fine for now.

Please take note of a comment on line 556 in models.py.

Also, I messed up the migration file 0037.  I reverted back to the previous migration to make some changes, but I see there were things in there about migrating stream data and the function 'copy_next_prompt_to_next_button_for_message'

Did you add those manually?

File: home/migrations/0037_alter_contentpage_whatsapp_body.py [hunk 1]
```diff
@@ -76,6 +75,24 @@ class Migration(migrations.Migration):
...
+                                                        label="Example Value"
+                                                    ),
+                                                )
+                                            ]
+                                        ),
+                                        default=[],
+                                        help_text="Please add example values for all variables used in a WhatsApp template",
+                                        label="Variable Example Values",
+                                    ),
+                                ),
```
Comment (rudigiesler) [inline home/migrations/0037_alter_contentpage_whatsapp_body.py:95]: For this migration, we probably don't want to edit this existing migration, because systems that already have this migration applied won't then do this additional step. What I would do here is revert the changes to this file, and run `makemigrations` again, to generate this as an `0038` migration.

File: home/models.py [hunk 2]
```diff
@@ -193,6 +199,12 @@ class WhatsappBlock(blocks.StructBlock):
         "media cannot exceed 1024 characters.",
         validators=(MaxLengthValidator(4096),),
     )
+    example_values = blocks.ListBlock(
+        ExampleValuesBlock(),
```
Comment (rudigiesler) [inline home/models.py:203]: Do we need the children here to be `StructBlock`s? Could we not just do a `blocks.ListBlock(blocks.CharBlock(....))`?

File: home/models.py [hunk 3]
```diff
@@ -212,19 +224,25 @@ class Meta:
 
     def clean(self, value):
         result = super().clean(value)
+        num_vars_in_msg = result["message"].count("{{")
```
Comment (rudigiesler) [inline home/models.py:227]: We could maybe use regex here to do a bit of a better match. ie. `len(re.findall("{{\d+}}", result["message"]))`

File: home/models.py [hunk 4]
```diff
@@ -212,19 +224,25 @@ class Meta:
...
+        errors = {}
+        if num_vars_in_msg > 0:
+            num_example_values = len(result["example_values"])
+            if num_vars_in_msg != num_example_values:
+                errors["example_values"] = ValidationError(
+                    "The number of example values provided does not match the number of variables used in the template"
```
Comment (rudigiesler) [inline home/models.py:233]: Maybe we could add the number of example values provided, and the number of variables that we counted in the message, to this error message? eg. `The number of example values provided (2) does not match the number of variables used in the template (3)`

File: home/models.py [hunk 5]
```diff
@@ -532,6 +551,20 @@ def whatsapp_template_body(self):
     def whatsapp_template_image(self):
         return self.whatsapp_body.raw_data[0]["value"]["image"]
 
+    @property
+    def whatsapp_template_example_values(self):
+        # TO CHECK IN REVIEW - I know there must be a cleaner way of doing this, but I can't seem to come up with it
```
Comment (rudigiesler) [inline home/models.py:556]: Maybe a:
```
example_values = self.whatsapp_body.raw_data[0]["value"].get("example_values", [])
return [v["value"]["example_values"] for v in example_values]
````

File: home/tests/test_content_import_export.py [hunk 6]
```diff
@@ -327,6 +327,14 @@ def remove_buttons(page: DbDict) -> DbDict:
     return page
 
 
+@per_page
+def remove_example_values(page: DbDict) -> DbDict:
+    if "whatsapp_body" in page["fields"]:
+        for body in page["fields"]["whatsapp_body"]:
+            body["value"].pop("example_values", None)
+    return page
```
Comment (rudigiesler) [inline home/tests/test_content_import_export.py:335]: We probably want to add the example values to the export and import, but happy for that to happen in another PR

File: home/models.py (hunk 2 above)
Comment (fritzbrand) [reply home/models.py:203]: 🤦 Thanks, fixed

File: home/migrations/0037_alter_contentpage_whatsapp_body.py (hunk 1 above)
Comment (fritzbrand) [reply home/migrations/0037_alter_contentpage_whatsapp_body.py:95]: Cool thanks

File: home/models.py (hunk 3 above)
Comment (fritzbrand) [reply home/models.py:227]: Cool, thanks

File: home/models.py (hunk 4 above)
Comment (fritzbrand) [reply home/models.py:233]: Cool, done

File: home/models.py (hunk 5 above)
Comment (fritzbrand) [reply home/models.py:556]: Rad, thanks.  I modified it slightly as the structure changed when I removed the StructBlocks

File: home/tests/test_content_import_export.py (hunk 6 above)
Comment (fritzbrand) [reply home/tests/test_content_import_export.py:335]: Cool, will do

File: home/migrations/0037_alter_contentpage_whatsapp_body.py (hunk 1 above)
Comment (rudigiesler) [reply home/migrations/0037_alter_contentpage_whatsapp_body.py:95]: I think we still need to remove these changes here

File: home/whatsapp.py [hunk 7]
```diff
@@ -40,6 +54,7 @@ def create_whatsapp_template(name, body, quick_replies=(), image_id=None):
         "language": "en_US",
         "components": components,
     }
+    print(json.dumps(data, indent=4))
```
Comment (rudigiesler) [inline home/whatsapp.py:57]: We can probably remove this print?

**Review by rudigiesler — COMMENTED** [review]
Looks good! Just some small things, looks like the changes to the existing migration are still included in the PR

File: home/migrations/0037_alter_contentpage_whatsapp_body.py (hunk 1 above)
Comment (fritzbrand) [reply home/migrations/0037_alter_contentpage_whatsapp_body.py:95]: Sorry, I discarded the changes in git, but looks like that didn't work.  Tried copying down the contents of the file that was in main now

File: home/whatsapp.py (hunk 7 above)
Comment (fritzbrand) [reply home/whatsapp.py]: Yup, sorry it snuck back in when I removed the StructBlocks and had to check what that looked like in the request

**Review by rudigiesler — APPROVED** [review]
Looks good!
